### DISCUSSION TITLE
CGAL Surface mesh python bindings

### DISCUSSION THREAD

**Discussion by darikg** [pyvista/pyvista#2446] [Show and tell]
Hi everybody,

I've been slowly picking away at writing python bindings for CGAL's [Surface Mesh](https://doc.cgal.org/latest/Surface_mesh/index.html) and [Polygon Mesh Processing](https://doc.cgal.org/latest/Polygon_mesh_processing/index.html) modules. I've finally gotten to a point where an extremely limited pre-alpha version is available to share publicly. I'm posting it here first because 1) the API is heavily inspired by pyvista and 2) I think the feature set complements pyvista nicely, filling in some stuff that isn't possible (AFAIK) in VTK, especially:

  - more robust boolean operations
  - remeshing/smoothing/refining/fairing on specific vertex/face subsets rather than the entire mesh
  - surface parameterization
  - geodesics between arbitrary face barycentric coordinates rather than just vertices
  - the heat method for fast geodesic distance approximation

if anyone finds any of this useful, contributions are more than welcome!

edit: forgot to add the link: [Seagull Mesh](https://github.com/darikg/seagullmesh)

**Comment by akaszynski** [comment]
Thanks for posting! I'd love to be able to add a link to your library, especially for the boolean operations.

**Reply by darikg** [reply]
Sure!

**Reply by akaszynski** [reply]
> I've finally gotten to a point where an extremely limited pre-alpha version is available to share publicly.

Would you be able to share the new library?

**Reply by akaszynski** [reply]
Actually, is this it?
https://github.com/darikg/seagullmesh

**Reply by darikg** [reply]
Oh wow I actually forgot to put the link in the original comment, that's embarassing. Just edited the original post

**Reply by akaszynski** [reply]
> Oh wow I actually forgot to put the link in the original comment, that's embarassing. Just edited the original post

Don't worry about it; that's what google is for.

One comment: would be awesome to not need conda for this. I realize that the conda package manager is a great way to get these sorts of packages, but I (personally) still prefer pip. If it's out of scope, I understand.

**Reply by darikg** [reply]
Totally agreed -- it's on the todo list

**Reply by whophil** [reply]
Took a (very) quick look and it seems like this is a pybind11 project. I don't know if this approach is feasible for CGAL, but here are examples of two separate projects which provide auto-generated binding code for another large C++ library, OpenCascade.

- https://github.com/CadQuery/OCP
- https://github.com/trelau/pyOCCT

Both projects contain (or reference via submodule) code which parses headers and generate corresponding pybind11 source files. I believe both packages using Clang for the parsing. I don't think either project is in a state yet where the binding generation code can be applied to another library without some serious modification, but perhaps there are some ideas that could be of interest to you.

Just thought I'd mention this, as writing pybind11 code for large libraries can be tedious and error-prone.

**Reply by darikg** [reply]
There's also [CGAL-swig-bindings](https://github.com/CGAL/cgal-swig-bindings) which I believe is auto-generated. My experience trying to work with it is that it feels like writing C++ in python, whereas I wanted something more pythonic and high-level, where it seems like hand-crafting bindings is inescapable. No argument it's tedious and error-prone though.

**Reply by whophil** [reply]
@darikg using auto-generated bindings will certainly lead to writing C++-like code in Python.

PyVista makes the "auto-generated" VTK bindings more Pythonic, but does this entirely on the Python side, rather than in the language of the binding code. Generally, I prefer this approach because: 
1. features you might want Seagull to cover in the future can be implemented in pure Python code
2. users/developers can fall back to the complete set of CGAL bindings if necessary (just like PyVista and VTK)

Not a criticism at all, just wanted to share some thoughts in the early stage of your project. Thanks for sharing!

**Reply by darikg** [reply]
good points, appreciate your thoughts!

**Comment by stla** [comment]
Cool. I wrote a R package which wraps some features of CGAL, mainly surface reconstruction: <https://laustep.github.io/stlahblog/posts/SurfaceReconstruction.html>. I'm not very skilled in C/C++ and I find hard to understand CGAL. Maybe your code will be more understandable for me, and if so I will try to implement more features :-)

**Reply by darikg** [reply]
Nice! I know [pygalmesh](https://github.com/meshpro/pygalmesh/) has also wrapped a bunch of CGAL's surface reconstruction methods.

I'm also not a very knowledgable C++ programmer so take everything in my code with a grain of salt

**Reply by stla** [reply]
Thanks. No problem. R CMD CHECK is very good for complaining about bad code :). I also use CPPCHECK but I have difficulties to make it working well with **Rcpp** (the R package we use to include C++ code, maybe comparable with Cython for Python). 
I already took a look at your code, and yes, it is helpful for me. Maybe I will start to plagiate it today :-)

**Comment by stla** [comment]
Hello @darikg 

Did you implement the Boolean operations? This is amazing. However I have an issue with one example. Maybe you can try it, if you did the implementation.

![Difference](https://user-images.githubusercontent.com/4466543/167486412-fcafbbdd-50c0-4632-b972-d8653a599ab0.png)

![Intersection](https://user-images.githubusercontent.com/4466543/167486429-5b7799e3-de6b-4855-b154-c4f759ad23de.png)

![Union](https://user-images.githubusercontent.com/4466543/167486443-86d36858-98ea-428e-92d4-4c74af0777a1.png)

**Reply by darikg** [reply]
Do you have sample code? And what's the issue?

**Reply by stla** [reply]
I have an issue with the compound of five tetrahedra. When I compute the intersection of the five tetrahedra, this works fine *if I round the vertex coordinates* but if I don't round them this crashes. If you implemented the intersection and if you're ok for trying, please say me how do you enter the vertex coordinates and the faces (numpy arrays?), so that I can prepare the Python code for you. 

![tetrahedraCompound](https://user-images.githubusercontent.com/4466543/167923957-eddb708f-e0bd-41de-9013-ffd006341604.gif)

**Reply by darikg** [reply]
cool, if you give me a quick pyvista script to generate the tetrahedra I'll try it in cgal

**Reply by stla** [reply]
Thanks. Code below.

```python
import numpy as np
import pyvista as pv

# all vertices ####
phi = (1 + np.sqrt(5)) / 2
a = 1 / np.sqrt(3)
b = a / phi
c = a * phi

vertices = np.vstack(
    (
        np.array([a, a, a]),
        np.array([a, a, -a]),
        np.array([a, -a, a]),
        np.array([-a, -a, a]),
        np.array([-a, a, -a]),
        np.array([-a, a, a]),
        np.array([0, b, -c]),
        np.array([0, -b, -c]),
        np.array([0, -b, c]),
        np.array([c, 0, -b]),
        np.array([-c, 0, -b]),
        np.array([-c, 0, b]),
        np.array([b, c, 0]),
        np.array([b, -c, 0]),
        np.array([-b, -c, 0]),
        np.array([-b, c, 0]),
        np.array([0, b, c]),
        np.array([a, -a, -a]),
        np.array([c, 0, b]),
        np.array([-a, -a, -a]),
    )
)

# tetrahedra vertices indices ####
tetra1Idxs = [16, 13, 1, 10]
tetra2Idxs = [17, 0, 3, 4]
tetra3Idxs = [18, 5, 14, 6]
tetra4Idxs = [2, 12, 11, 7]
tetra5Idxs = [19, 15, 9, 8]

# tetrahedra meshes ####
faces = np.array([3, 0, 1, 2, 3, 2, 1, 3, 3, 3, 1, 0, 3, 0, 2, 3])
th1 = pv.PolyData(vertices[tetra1Idxs], faces)
th2 = pv.PolyData(vertices[tetra2Idxs], faces)
th3 = pv.PolyData(vertices[tetra3Idxs], faces)
th4 = pv.PolyData(vertices[tetra4Idxs], faces)
th5 = pv.PolyData(vertices[tetra5Idxs], faces)
```

**Reply by darikg** [reply]
Thanks! There's indeed an issue. Moving the discussion to https://github.com/darikg/seagullmesh/issues/1
